compiler/rustc_hir_typeck/src/expr.rs

```diff
@@ -1746,7 +1746,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                                 let target_ty = self.field_ty(base_expr.span, f, args);
                                 let cause = self.misc(base_expr.span);
                                 match self.at(&cause, self.param_env).sup(
-                                    DefineOpaqueTypes::No,
+                                    DefineOpaqueTypes::Yes,
```

Comment:
This needs `#![feature(type_changing_struct_update)]`. I think a test could be crafted using that.